src/orange/surf/SoftSurfaceEqual.cc

```diff
@@ -0,0 +1,173 @@
+#include "SoftSurfaceEqual.hh"
+
+#include <cmath>
+
+#include "detail/AllSurfaces.hh"
+
+namespace celeritas
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+#define ORANGE_INSTANTIATE_OP(S)                                         \
+    template bool SoftSurfaceEqual::operator()(S<Axis::x> const&,        \
+                                               S<Axis::x> const&) const; \
+    template bool SoftSurfaceEqual::operator()(S<Axis::y> const&,        \
+                                               S<Axis::y> const&) const; \
+    template bool SoftSurfaceEqual::operator()(S<Axis::z> const&,        \
+                                               S<Axis::z> const&) const
+
+//---------------------------------------------------------------------------//
+}  // namespace
+
+//---------------------------------------------------------------------------//
+/*!
+ * Compare two aligned planes for near equality.
+ */
+template<Axis T>
+bool SoftSurfaceEqual::operator()(PlaneAligned<T> const& a,
+                                  PlaneAligned<T> const& b) const
+{
+    return soft_eq_(a.position(), b.position());
+}
+
+//! \cond
+ORANGE_INSTANTIATE_OP(PlaneAligned);
+//! \endcond
+
+//---------------------------------------------------------------------------//
+/*!
+ * Compare two centered axis-aligned cylinders for near equality.
+ */
+template<Axis T>
+bool SoftSurfaceEqual::operator()(CylCentered<T> const& a,
+                                  CylCentered<T> const& b) const
+{
+    return soft_eq_sq_(a.radius_sq(), b.radius_sq());
+}
+
+//! \cond
+ORANGE_INSTANTIATE_OP(CylCentered);
+//! \endcond
+
+//---------------------------------------------------------------------------//
+/*!
+ * Compare two centered spheres for near equality.
+ */
+bool SoftSurfaceEqual::operator()(SphereCentered const& a,
+                                  SphereCentered const& b) const
+{
+    return soft_eq_sq_(a.radius_sq(), b.radius_sq());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Compare two aligned cylinders for near equality.
+ */
+template<Axis T>
+bool SoftSurfaceEqual::operator()(CylAligned<T> const& a,
+                                  CylAligned<T> const& b) const
+{
+    return soft_eq_sq_(a.radius_sq(), b.radius_sq())
+           && soft_eq_distance_(a.calc_origin(), b.calc_origin());
+}
+
+//! \cond
+ORANGE_INSTANTIATE_OP(CylAligned);
+//! \endcond
+
+//---------------------------------------------------------------------------//
+/*!
+ * Compare two planes for near equality.
+ */
+bool SoftSurfaceEqual::operator()(Plane const& a, Plane const& b) const
+{
+    // Guard against dot product being slightly greater than 1 due to fp
+    // arithmetic
+    real_type const ndot = dot_product(a.normal(), b.normal());
+    return (ndot >= 1 || std::sqrt(1 - ndot) < soft_eq_.rel() * 3)
```

Comment:
I don't understand what is going on here with `std::sqrt(1 - ndot)`. I am thinking in terms of $a \cdot b = |a| |b| cos(\theta)$. With $a$ and $b$ as normalized vectors, we get  $\sqrt(1 - cos(\theta))$ which I don't think has any physical significance. `std::sqrt(1 - ndot*ndot)` would be $\sin(\theta)$, but I don't think that makes sense either. What about the factor of 3?

Reply:
You're right, that definition is ad hoc. (The 3  comes from the new `is_soft_unit_vector`; see https://github.com/celeritas-project/celeritas/pull/932/files#diff-aed6792f4adb3b016a019a1f65191218845f3e9aebdeb4e226b04c751c17f971 .) I've replaced it from a fresh derivation that ensures that the error δ in intercept to two slightly different planes is less than the tolerance ε:
<img width="444" alt="Screenshot 2023-09-12 at 08 11 43" src="https://github.com/celeritas-project/celeritas/assets/741229/b2bbcf53-dbb6-40fa-b108-13b1996024db">

Reply:
This is a great diagram to accompany your description in the doc string, is there someway it could be included and referenced there?

Reply:
Maybe it'll go in the ORANGE paper 😅